### PR TITLE
improve installer tests compatibility with pit

### PR DESCRIPTION
**Pull request by rmynar**
This change improves compatibility with PIT testing.

- added RHEL_SOURCE variable to robottelo.conf to choose which repositories to use. ga - use released RHEL, internal - use custom RHEL (e.g. release candidate)
- removed 'beta' from available sources of satellite and capsule as we don't use beta releases any more
- updated validators

We can choose source repositories for RHEL and Satellite. Following combinations of sat@rhel are possible:
internal@internal - pre-release testing
ga@internal - PIT testing
internal@ga - common testing
ga@ga - invalid combination (we don't care about retesting already published products)
The same applies for Capsule. It's also possible to use different version of Capsule and Satellite (e.g. n-1 testing)

- reworked satellite installation - use RHEL and Satellite sources as defined by _robottelo.rhel_source_ and _server.version_
- reworked capsule installation - setup RHEL and Capsule sources on Satellite, register capsule as host and perform capsule installation

See SAT-26896

### CODE REVIEW

**Comment by ogajduse** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7403
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ============ 1 failed, 256 warnings, 1 error in 2759.16s (0:45:59) =============
```

**Comment by rmynar** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7419
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================= 2 passed, 404 warnings in 4529.02s (1:15:29) =================
```

File: tests/foreman/installer/test_installer.py [hunk 1]
```diff
@@ -195,10 +197,10 @@ def test_capsule_installation(sat_non_default_install, cap_ready_rhel, setting_u
     :id: 64fa85b6-96e6-4fea-bea4-a30539d59e65
 
     :steps:
-        1. Get a fapolicyd enabled Satellite
-        2. Configure capsule repos
-        3. Install and enable fapolicyd
-        4. Enable capsule module
+        1. Use Satellite with fapolicyd enabled (non-default)
```
Comment (ekohl) [inline tests/foreman/installer/test_installer.py:200]: I see this uses `sat_non_default_install` but I really disagree that we should be testing with `--foreman-rails-cache-store type:file`. If we want proper integration testing we should be testing a basic default Satellite instead of the edge case.

If anything, the test that actually wants to test Satellite with non-Redis should explicitly configure Satellite that way.

**Review by ekohl — COMMENTED** [review]
Structure wise I it feels like it does a lot. Would it make sense to split it up into a helper? Like a fixture that creates the product and repositories. Then create an activation key for those.

File: pytest_fixtures/core/sat_cap_factory.py [hunk 2]
```diff
@@ -358,13 +359,28 @@ def installer_satellite(request):
...
+    # register to cdn (also enables rhel repos from cdn)
     sat.register_to_cdn(enable_proxy=True)
-    sat.download_repofile(
-        product='satellite',
-        release=settings.server.version.release,
-        snap=settings.server.version.snap,
-    )
+
+    # setup source repositories
+    if settings.server.version.source == "ga":
```
Comment (lpramuk) [inline pytest_fixtures/core/sat_cap_factory.py:367]: `"ga"` ?!??
Wouldn't be `"cdn"` better ?

File: tests/foreman/installer/test_installer.py [hunk 3]
```diff
@@ -212,25 +214,155 @@ def test_capsule_installation(sat_non_default_install, cap_ready_rhel, setting_u
...
+            ak.content_override(
+                data={
+                    'content_overrides': [
+                        {
+                            'content_label': '_'.join([org.label, product_rhel.label, repo.label]),
+                            'value': '1',
+                        }
+                    ]
+                }
+            )
```
[the remark this answers is not in the document]
Comment (rmynar) [reply tests/foreman/installer/test_installer.py:183]: I moved the steps to prepare the repositories into separate function, so it can be re-used in other places. But I think it's not necessary to move it outside of this test suite, because this is the only place where we observe the installation process. (Other tests use templates with pre-installed Satellite/Capsule) .
Luckily the content override works "one by one", so all the repositories are enabled at the end. Performance impact is negligible - the API call takes just few milliseconds.

File: tests/foreman/installer/test_installer.py [hunk 4]
```diff
@@ -212,25 +214,155 @@ def test_capsule_installation(sat_non_default_install, cap_ready_rhel, setting_u
...
+        )
+
+    # Wait for asynchronous sync tasks
+    logger.info(f'Waiting for {len(sync_tasks)} sync tasks')
+    for task in sync_tasks:
+        logger.info(f'Waiting for task {task['id']}')
+        sat_non_default_install.wait_for_tasks(
+            search_query=(f'id = {task["id"]}'),
+            poll_timeout=1800,
+        )
```
[the remark this answers is not in the document]
Comment (rmynar) [reply tests/foreman/installer/test_installer.py]: You're right, the function has its own loop so it can accept multiple tasks in search query.

File: tests/foreman/installer/test_installer.py (hunk 1 above)
Comment (rmynar) [reply tests/foreman/installer/test_installer.py]: I decided to create another fixture with default setup but including fapolicyd. Let's consider this as a workaround I hope it can be done better (e.g. parametrized), but it's out of scope of this PR.

File: pytest_fixtures/core/sat_cap_factory.py (hunk 2 above)
Comment (rmynar) [reply pytest_fixtures/core/sat_cap_factory.py:359]: This is a remainder of historical state when there were two versions available on cdn - ga and beta. Beta is no more used. However it still seems logical to me - we have internal source or generally available source. In case of renaming I would suggest different naming`internal`/`public`.

File: tests/foreman/installer/test_installer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekohl) [reply tests/foreman/installer/test_installer.py:183]: I don't know the robottelo code base well (but am familiar with pytest). Is this something you'd use as a fixture? The reuse is probably easier, but I can also see some downsides in debugging (heavy fixtures that fail are often a pain to debug).

**Review by ekohl — APPROVED** [review]
With my limited robottelo knowledge, it looks good to me.

**Comment by rmynar** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7851
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================= 239 warnings, 2 errors in 2373.01s (0:39:33) =================
```

**Comment by rmynar** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7860
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================= 2 passed, 454 warnings in 4625.61s (1:17:05) =================
```

**Review by ogajduse — APPROVED** [review]

**Review by jameerpathan111 — APPROVED** [review]

File: tests/foreman/installer/test_installer.py [hunk 5]
```diff
@@ -165,6 +273,19 @@ def sat_default_install(module_sat_ready_rhels):
     return sat
 
 
+@pytest.fixture(scope='module')
+def sat_default_install_fapolicyd(module_sat_ready_rhels):
```
Comment (jameerpathan111) [inline tests/foreman/installer/test_installer.py:277]: @rmynar `sat_default_install` and `sat_default_install_fapolicyd` are both using `module_sat_ready_rhels` won't this affect tests in some way?

File: conf/server.yaml.template [hunk 6]
```diff
@@ -4,15 +4,15 @@ SERVER:
...
+    RELEASE: 6.15.0
     # The snap version currently testing (if applicable)
     SNAP: 1.0
     # The source of Satellite packages. Can be one of:
-    # internal, ga, beta
+    # internal, ga
     SOURCE: "internal"
     # The RHEL Base OS Version(x.y) where the Satellite is installed
-    RHEL_VERSION: '7'
+    RHEL_VERSION: '8'
```
Comment (jameerpathan111) [inline conf/server.yaml.template:15]: Why not RHEL9 at least for master/6/16/0?

File: pytest_fixtures/core/sat_cap_factory.py [hunk 7]
```diff
@@ -348,13 +349,28 @@ def installer_satellite(request):
...
+            sat.enable_repo(repo, force=True)
+    else:
+        # get ohsnap repofile
+        sat.download_repofile(
+            product='satellite',
+            release=settings.server.version.release,
+            snap=settings.server.version.snap,
+        )
+    if settings.robottelo.rhel_source == "internal":
+        # disable rhel repos from cdn
```
Comment (jameerpathan111) [inline pytest_fixtures/core/sat_cap_factory.py:369]: Wouldn't `if elif` be better here instead of trying to disable repos later?

File: robottelo/host_helpers/contenthost_mixins.py [hunk 8]
```diff
@@ -56,6 +56,22 @@ def REPOS(self):
...
+    def SATELLITE_CDN_REPOS(self):
+        sat_version = ".".join(settings.server.version.release.split('.')[0:2])
+        return {
+            'satellite': f"satellite-{sat_version}-for-rhel-{self._v_major}-x86_64-rpms",
+            'sat-maintenance': f"satellite-maintenance-{sat_version}-for-rhel-{self._v_major}-x86_64-rpms",
+        }
+
+    @cached_property
+    def CAPSULE_CDN_REPOS(self):
+        sat_version = ".".join(settings.server.version.release.split('.')[0:2])
```
Comment (jameerpathan111) [inline robottelo/host_helpers/contenthost_mixins.py:69]: `sat_version` could be made a constant to avoid code duplication in this module.

File: robottelo/host_helpers/contenthost_mixins.py [hunk 9]
```diff
@@ -56,6 +56,22 @@ def REPOS(self):
...
+        return {
+            'capsule': f"satellite-capsule-{sat_version}-for-rhel-{self._v_major}-x86_64-rpms",
```
Comment (jameerpathan111) [inline robottelo/host_helpers/contenthost_mixins.py:71]: I'd suggest we make use of robottelo constants here. Probably make the existing one more robust like you have done here, something which need not be hard coded for each version. https://github.com/SatelliteQE/robottelo/blob/fa9126302531728fd602ff7b895f59b93814f567/robottelo/constants/__init__.py#L371

**Review by jameerpathan111 — CHANGES_REQUESTED** [review]

**Comment by rmynar** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8067
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================== 7 warnings, 2 errors in 208.09s (0:03:28) ===================
```

**Comment by rmynar** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8076
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================== 7 warnings, 2 errors in 206.69s (0:03:26) ===================
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8077
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================== 7 warnings, 2 errors in 202.26s (0:03:22) ===================
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8079
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================= 2 passed, 522 warnings in 4548.05s (1:15:48) =================
```

**Comment by jameerpathan111** [comment]
@rmynar I converted it to draft for now, please mark it ready for review once you're done with changes.

**Comment by ogajduse** [comment]
trigger: test-robottelo
pytest: tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8310
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/installer/test_installer.py::test_satellite_installation tests/foreman/installer/test_installer.py::test_capsule_installation --external-logging
Test Result : ================= 2 passed, 519 warnings in 4686.80s (1:18:06) =================
```

File: pytest_fixtures/core/sat_cap_factory.py (hunk 7 above)
Comment (ogajduse) [reply pytest_fixtures/core/sat_cap_factory.py:369]: `sat.register_to_cdn(enable_proxy=True)` enables AppStream and BaseOS, therefore in case we want to consume Satellite content from CDN, but we want to consume RHEL content from our internal source, we need to disable these two repos.

File: robottelo/host_helpers/contenthost_mixins.py (hunk 8 above)
Comment (ogajduse) [reply robottelo/host_helpers/contenthost_mixins.py:69]: @jameerpathan111 Can you be more specific? Where would that constant be placed and where should it take the version data from?

File: robottelo/host_helpers/contenthost_mixins.py (hunk 9 above)
Comment (ogajduse) [reply robottelo/host_helpers/contenthost_mixins.py:71]: To me, constants are static, we need to expand them once a new version of <something> gets released. Contrary to that, what Radek proposes is a dynamic approach.
@jameerpathan111 How are these two cached_properties different from the above methods/properties? Is there something you would like to improve on the above methods as well?

Comment (ogajduse) [reply robottelo/host_helpers/contenthost_mixins.py:71]: Looking at the methods above, I see what you mean. Do you see any danger in keeping what Radek proposes?

File: tests/foreman/installer/test_installer.py (hunk 5 above)
Comment (jyejare) [reply tests/foreman/installer/test_installer.py]: @jameerpathan111 They might be using two different satellites though cause `module_sat_ready_rhels` is spinning 2 satellites.

TBH, this is a very bad design of spinning two sats and deciding them in 2 tests. Even though you wish to run just 1 test out of it, the fixture will unnecessarily spin up 2 systems.

I am changing that in https://github.com/SatelliteQE/robottelo/pull/15948/files#diff-606dfd8ed66a37a0a150b787d8ce417b246c8d3e0aaf8b33c4d35da5189f5792R161

File: robottelo/host_helpers/contenthost_mixins.py (hunk 9 above)
Comment (jameerpathan111) [reply robottelo/host_helpers/contenthost_mixins.py:71]: I don't see any danger, it was just a suggestion.

File: robottelo/host_helpers/contenthost_mixins.py (hunk 8 above)
Comment (jameerpathan111) [reply robottelo/host_helpers/contenthost_mixins.py:69]: In `class VersionedContent`, to reduce code duplication as both `CAPSULE_CDN_REPOS` and `SATELLITE_CDN_REPOS` use it.

Comment (ogajduse) [reply robottelo/host_helpers/contenthost_mixins.py:69]: I will leave this one up to Radek to decide.

File: pytest_fixtures/core/sat_cap_factory.py (hunk 7 above)
Comment (jameerpathan111) [reply pytest_fixtures/core/sat_cap_factory.py:369]: ah ok. I get it.

File: conf/server.yaml.template (hunk 6 above)
Comment (rmynar) [reply conf/server.yaml.template]: it's just a template ;-) but I bumped the version :-)

File: tests/foreman/installer/test_installer.py (hunk 5 above)
Comment (rmynar) [reply tests/foreman/installer/test_installer.py]: If I understand correctly the original intent of this implementation, the purpose was to make testing faster. All hosts are spinned up concurrently during setup, then tests are run sequentially.
Btw i'm only changing the count of deployed hosts, not the implementation itself.

File: robottelo/host_helpers/contenthost_mixins.py (hunk 9 above)
Comment (rmynar) [reply robottelo/host_helpers/contenthost_mixins.py:71]: There is already kind of sat_version constant in satellite object, however here we can't use satellite object as we only have a clean RHEL host where we're going to install Satellite. Also there is already a configuration file containing repository ids, but it does not get updated within PIT tests. These are the reasons why I came up with this solution.
Thanks for calling it "robust". We'll see how it works and possibly we could start using it more often. But currently (wthin this PR) I don't want to use it outside of Installer tests.

**Review by jameerpathan111 — APPROVED** [review]
